format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/PathTrackingDelimitedInputFormat.java

```diff
@@ -69,66 +69,68 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
         if (delegate.nextKeyValue()) {
           // skip to next if the current record is header
           if (skipHeader && delegate.getCurrentKey().get() == 0L) {
-            return delegate.nextKeyValue();
+            if (!delegate.nextKeyValue()) {
+              return false;
+            }
+          }
+          // getCurrentValue logic moved to nextKeyValue to avoid discrepancies in the data if there are multiple
+          // calls to getCurrentValue method.
+          String delimitedString = delegate.getCurrentValue().toString();
+          builder = StructuredRecord.builder(schema);
+          Iterator<Schema.Field> fields = schema.getFields().iterator();
+          Iterator<String> splitsIterator = getSplitsIterator(enableQuotesValue, delimitedString, delimiter);
+          int dataFieldsCount = 0;
+          while (splitsIterator.hasNext()) {
+            dataFieldsCount++;
+            String part = splitsIterator.next();
+            if (!fields.hasNext()) {
+              handleImproperString(delimitedString.contains(QUOTE), dataFieldsCount);
+            }
+
+            Schema.Field nextField = fields.next();
+            DelimitedStructuredRecordStringConverter.parseAndSetFieldValue(builder, nextField, part);
           }
           return true;
         }
         return false;
       }
 
+      private void handleImproperString(boolean containsQuote, int numDataFields) throws IOException {
+        int numSchemaFields = schema.getFields().size();
+        String message =
+          String.format(
+            "Found a row with %d fields when the schema only contains %d field%s.",
+            numDataFields, numSchemaFields, numSchemaFields == 1 ? "" : "s");
```

Comment:
this gets called when the schema has fewer fields than the line, so numDataFields will be equal to numSchemaFields here. Before this, need to finish iterating through the splitsIterator in order to get the real number of data fields.